File: sdks/python/meerkat/event_envelope.py
"""Strict parsing for canonical agent-event envelopes."""

from __future__ import annotations

from typing import Any
from uuid import UUID

from .errors import MeerkatError
from .events import parse_event
from .types import EventEnvelope, EventSourceIdentity

MAX_U64 = (1 << 64) - 1


def _invalid(reason: str, raw: Any) -> MeerkatError:
    return MeerkatError(
        "INVALID_RESPONSE",
        f"Invalid agent event envelope: {reason}",
        details=raw,
    )


def _require_record(raw: Any, field: str, envelope: Any) -> dict[str, Any]:
    if not isinstance(raw, dict):
        raise _invalid(f"{field} must be an object", envelope)
    return raw


def _require_non_empty_string(raw: dict[str, Any], field: str, envelope: Any) -> str:
    value = raw.get(field)
    if not isinstance(value, str) or not value:
        raise _invalid(f"{field} must be a non-empty string", envelope)
    return value


def _require_uuid(raw: dict[str, Any], field: str, envelope: Any) -> str:
    value = _require_non_empty_string(raw, field, envelope)
    try:
        parsed = UUID(value)
    except ValueError as error:
        raise _invalid(f"{field} must be a canonical UUID", envelope) from error
    if str(parsed) != value:
        raise _invalid(f"{field} must be a canonical UUID", envelope)
    return value


def _require_unsigned_integer(raw: dict[str, Any], field: str, envelope: Any) -> int:
    value = raw.get(field)
    if (
        isinstance(value, bool)
        or not isinstance(value, int)
        or value < 0
        or value > MAX_U64
    ):
        raise _invalid(f"{field} must be an unsigned 64-bit integer", envelope)
    return value


def parse_event_source_identity(raw: Any, envelope: Any) -> EventSourceIdentity:
    """Parse the required discriminated event-source identity."""

    record = _require_record(raw, "source", envelope)
    source_type = _require_non_empty_string(record, "type", envelope)
    if source_type == "session":
        return EventSourceIdentity(
            type=source_type,
            session_id=_require_uuid(record, "session_id", envelope),
        )
    if source_type == "runtime":
        return EventSourceIdentity(
            type=source_type,
            runtime_id=_require_non_empty_string(record, "runtime_id", envelope),
        )
    if source_type == "interaction":
        return EventSourceIdentity(
            type=source_type,
            interaction_id=_require_uuid(record, "interaction_id", envelope),
        )
    if source_type == "callback":
        return EventSourceIdentity(type=source_type)
    if source_type == "external":
        return EventSourceIdentity(
            type=source_type,
            source_id=_require_non_empty_string(record, "source_id", envelope),
        )
    raise _invalid(f"source.type has unsupported value {source_type!r}", envelope)


def parse_agent_event_envelope(raw: Any) -> EventEnvelope:
    """Parse a full wire envelope without fabricating identity/order facts."""

    record = _require_record(raw, "envelope", raw)
    payload = _require_record(record.get("payload"), "payload", raw)
    mob_id: str | None = None
    if record.get("mob_id") is not None:
        mob_id = _require_non_empty_string(record, "mob_id", raw)
    return EventEnvelope(
        event_id=_require_uuid(record, "event_id", raw),
        source=parse_event_source_identity(record.get("source"), raw),
        seq=_require_unsigned_integer(record, "seq", raw),
        timestamp_ms=_require_unsigned_integer(record, "timestamp_ms", raw),
        payload=parse_event(payload),
        mob_id=mob_id,
    )

File: sdks/typescript/src/event-envelope.ts
/** Strict parsing for the canonical agent-event envelope. */

import { parseCoreEvent } from "./events.js";
import { MeerkatError } from "./generated/errors.js";
import type { AgentEventEnvelope, EventSourceIdentity } from "./types.js";

const CANONICAL_UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

function invalidEnvelope(
  reason: string,
  raw: unknown,
): MeerkatError {
  return new MeerkatError(
    "INVALID_RESPONSE",
    `Invalid agent event envelope: ${reason}`,
    raw,
  );
}

function requireRecord(raw: unknown, field: string, envelope: unknown): Record<string, unknown> {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
    throw invalidEnvelope(`${field} must be an object`, envelope);
  }
  return raw as Record<string, unknown>;
}

function requireNonEmptyString(
  raw: Record<string, unknown>,
  field: string,
  envelope: unknown,
): string {
  const value = raw[field];
  if (typeof value !== "string" || value.length === 0) {
    throw invalidEnvelope(`${field} must be a non-empty string`, envelope);
  }
  return value;
}

function requireUnsignedInteger(
  raw: Record<string, unknown>,
  field: string,
  envelope: unknown,
): number {
  const value = raw[field];
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) {
    throw invalidEnvelope(`${field} must be a non-negative safe integer`, envelope);
  }
  return value;
}

function requireUuid(
  raw: Record<string, unknown>,
  field: string,
  envelope: unknown,
): string {
  const value = requireNonEmptyString(raw, field, envelope);
  if (!CANONICAL_UUID.test(value)) {
    throw invalidEnvelope(`${field} must be a canonical UUID`, envelope);
  }
  return value;
}

export function parseEventSourceIdentity(
  raw: unknown,
  envelope: unknown,
): EventSourceIdentity {
  const record = requireRecord(raw, "source", envelope);
  const type = requireNonEmptyString(record, "type", envelope);
  switch (type) {
    case "session":
      return {
        type,
        sessionId: requireUuid(record, "session_id", envelope),
      };
    case "runtime":
      return {
        type,
        runtimeId: requireNonEmptyString(record, "runtime_id", envelope),
      };
    case "interaction":
      return {
        type,
        interactionId: requireUuid(record, "interaction_id", envelope),
      };
    case "callback":
      return { type };
    case "external":
      return {
        type,
        sourceId: requireNonEmptyString(record, "source_id", envelope),
      };
    default:
      throw invalidEnvelope(`source.type has unsupported value ${JSON.stringify(type)}`, envelope);
  }
}

export function parseAgentEventEnvelope(raw: unknown): AgentEventEnvelope {
  const record = requireRecord(raw, "envelope", raw);
  const payload = requireRecord(record.payload, "payload", raw);
  let mobId: string | undefined;
  if (record.mob_id !== undefined && record.mob_id !== null) {
    mobId = requireNonEmptyString(record, "mob_id", raw);
  }
  return {
    eventId: requireUuid(record, "event_id", raw),
    source: parseEventSourceIdentity(record.source, raw),
    seq: requireUnsignedInteger(record, "seq", raw),
    timestampMs: requireUnsignedInteger(record, "timestamp_ms", raw),
    payload: parseCoreEvent(payload),
    ...(mobId !== undefined ? { mobId } : {}),
  };
}
